crates/lib/src/qpu/api.rs

```diff
@@ -91,74 +72,287 @@ impl From<String> for JobId {
 
-    job_execution_id
+    Ok(job_execution_id
         .map(JobId)
-        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution ID".into()))
+        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution ID".into()))?)
 }
 
 /// Fetch results from QPU job execution.
+///
+/// # Arguments
+/// * `job_id` - The [`JobId`] to retrieve results for.
+/// * `quantum_processor_id` - The quantum processor the job was run on.
+/// * `client` - The [`Qcs`] client to use.
+/// * `connection_strategy` - The [`ConnectionStrategy`] to use. If this is a
+///       [`ConnectionStrategy::EndpointId`] then direct access to that endpoint
+///       overrides the `quantum_processor_id` parameter.
 pub async fn retrieve_results(
     job_id: JobId,
-    job_target: &JobTarget,
+    quantum_processor_id: String,
     client: &Qcs,
-) -> Result<ControllerJobExecutionResult, GrpcClientError> {
+    execution_options: &ExecutionOptions,
+) -> Result<ControllerJobExecutionResult, QpuApiError> {
     #[cfg(feature = "tracing")]
-    tracing::debug!("retrieving job results for {} on {}", job_id, job_target,);
+    tracing::debug!(
+        "retrieving job results for {} on {} using options {:?}",
+        job_id,
+        quantum_processor_id,
+        execution_options,
+    );
 
     let request = GetControllerJobResultsRequest {
         job_execution_id: job_id.0,
-        target: Some(job_target.into()),
+        target: Some(
+            execution_options
+                .connection_strategy
+                .get_results_target(&quantum_processor_id),
+        ),
     };
 
-    let mut controller_client = match job_target {
-        JobTarget::EndpointId(endpoint_id) => {
-            client
-                .get_controller_client_with_endpoint_id(endpoint_id)
-                .await
-        }
-        JobTarget::QuantumProcessorId(quantum_processor_id) => {
-            client.get_controller_client(quantum_processor_id).await
-        }
-    }?;
+    let mut controller_client = execution_options
+        .connection_strategy
+        .get_controller_client(client, &quantum_processor_id)
+        .await?;
 
-    controller_client
+    Ok(controller_client
         .get_controller_job_results(request)
-        .await?
+        .await
+        .map_err(GrpcClientError::RequestFailed)?
         .into_inner()
         .result
-        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))
+        .ok_or_else(|| GrpcClientError::ResponseEmpty("Job Execution Results".into()))?)
+}
+
+/// Options avaialable when executing a job on a QPU.
+///
+/// Use [`Default`] to get a reasonable set of defaults, or start with [`ExecutionOptionsBuilder`]
+/// to build a custom set of options.
+#[derive(Builder, Clone, Debug, Default, PartialEq, Eq)]
+pub struct ExecutionOptions {
+    #[doc = "The [`ConnectionStrategy`] to use to establish a connection to the QPU."]
+    connection_strategy: ConnectionStrategy,
+}
+
+impl ExecutionOptions {
```

Comment:
```suggestion
impl ExecutionOptions {
  pub fn builder() -> ExecutionOptionsBuilder {
    ExecutionOptionsBuilder::default()
  }
```

A common pattern I've seen and used to make it more ergonomic to get at the builder